### DISCUSSION TITLE
Razor Pages vs Liquid templates

### DISCUSSION THREAD

**Discussion by dzoech** [OrchardCMS/OrchardCore#6964] [General]
Hello!
Could somebody elaborate on the fundamental differences between Razor Pages and Liquid templates? What are the implications of using either? Advantages, drawbacks?

**Comment by Piedone** [comment]
You can find a lot of details about these in their respective official documentation but to summarize:

- Razor **templates** (what I think your questions is more likely about, so not Razor Pages) are templates written in the Razor syntax, which is basically HTML + C# + some funny stuff. If you know C# and HTML then this won't be a big challenge. Also, you can do anything from Razor that you can from C# otherwise.
- Liquid is another templating engine and syntax, so it's basically interchangeable in its usage with Razor. The major difference is that Liquid is designed to be safe: You can't break the app from Liquid. So it's safe to let people write it from the admin UI, even when you're running a public SaaS. However, it's also much more limited than Razor, and it is it's own language that you have to learn (meaning Liquid itself, otherwise it's HTML still).
- Razor Pages are still Razor templates with the same syntax, the difference is how they're used: Basic Razor templates in Orchard are used as part of the MVC paradigm, being views. So when you have a Razor template it always corresponds to a controller action (and there are partials and shape templates but let's keep it simple for now). Razor Pages, however, are a paradigm different than MVC: Instead of creating controllers that work with models and pass them on to views for display (and whatever else) you define your pages full in a single Razor template each. (This gets more complicated of course, and Razor Pages also operate with models, but this is the basic idea.)

All in all, with Orchard you can create any template in Razor or Liquid. Furthermore, you can also write the whole frontend of the site in Razor Pages instead of using the built-in MVC frontend; this we call a "decoupled" setup ("full CMS" being where you use the MVC frontend Orchard provides, "headless" being that Orchard only serves as a backend and is accessed from a completely separate app via APIs).

**Reply by dzoech** [reply]
Many thanks for your fast and details answer -- really helps me to verify my assumptions. 

To summarize it: if I don't have requirements for a public SaaS, i.e. no templates created by the user, Liquid doesn't bring any specific benefits? Currently, there is no existing knowledge for any of the two templating engines, however, ASP.NET experience is present.

**Reply by Piedone** [reply]
In your case, I'd go with Razor then. Liquid brings the benefit of being another widely used templating language but if you can use Razor too, and your developers know ASP.NET anyway, then Razor is better. Do note that in some cases you still have to use Liquid to configure something from the admin UI, like string patterns. But these are mostly really simple cases.

**Reply by dzoech** [reply]
Perfect, thanks again!

**Reply by SViradiya-MarutiTech** [reply]
Any performance difference . I would say Fluid and Razor pages. Fluid is based on Liquid templates

**Reply by Piedone** [reply]
You'll need to benchmark your specific use case. My gut feeling is that in a 1-1 comparison Razor should be lower on computing resources, since it's compiled to C# vs Liquid being interpreted. However, the contribution to a request's resource usage and especially latency should be relatively small of either one, compared to I/O that tends to dominate it.
